[C-1595] Fix sign out

raymondjacobson opened the pull request:
### Description
Fix sign out / sign in behavior. I felt like a small sensible refactor was due here -- I made SignOnScreen and HomeScreen both rendered at the same time so we can control via navigation what screen you're on as a user -- I think this is better? But open to feedback.

Some other issues:
- sign out sagas connected twice :oops:
- wait for account wasn't taking into consideration IDLE state
- sign out wasn't doing a page navigation but implicily relying on account state to change, which makes it hard to track
- sign in sometimes clears state before animations finish so you can see some error messages at the tail end of signing in sometimes in dev mode



### Dragons

*Is there anything the reviewer should be on the lookout for? Are there any dangerous changes?*

### How Has This Been Tested?

*Please describe the tests that you ran to verify your changes. Provide repro instructions & any configuration.*

- sign in
- sign in & sign out & sign in
- sign in & sign out & sign in different account
- sign in & sign out & quit app & sign in
- sign in & quit app & open app


### How will this change be monitored?

*For features that are critical or could fail silently please describe the monitoring/alerting being added.*

### Feature Flags ###

*Are all new features properly feature flagged? Describe added feature flags.*

sliptype reviewed (approved):
This all looks great but I'm a little worried about rendering both the signonstack and homestack at the same time. Are the lineups for the feed trying to load while on the signon screen?

raymondjacobson commented:
> This all looks great but I'm a little worried about rendering both the signonstack and homestack at the same time. Are the lineups for the feed trying to load while on the signon screen?

They do wait for you to have an account it seems but fair callout. I can undo that part of the change, but I am worried that the implicit navigation is a bit error prone.

raymondjacobson commented:
> > This all looks great but I'm a little worried about rendering both the signonstack and homestack at the same time. Are the lineups for the feed trying to load while on the signon screen?
> 
> They do wait for you to have an account it seems but fair callout. I can undo that part of the change, but I am worried that the implicit navigation is a bit error prone. 

So I think if this is a real concern we should actually switch this to render no HomeStack or SignOnStack until account is in success or error state. wdyt?

dylanjeffers commented on packages/mobile/src/store/sign-out/sagas.ts:
```diff
@@ -6,24 +6,24 @@ import { audiusBackendInstance } from 'app/services/audius-backend-instance'
 
-    yield* put(setSearchHistory([]))
+    yield* put(resetAccount())
+    yield* put(clearHistory())
```
great add, thank you

raymondjacobson commented:
New changes, please re-review!
- After discussion, switched back to the "render either or" set up for the RootScreen (this is an improvement of what it was before whether sometimes both would be rendered)
- Cleaned up a lot of code around regaining reachability
- Added waitForRead to a few more places
- Fixed and tested C-1289, which clears the WebApp entropy on sign out

raymondjacobson commented on packages/common/src/services/audius-backend/AudiusBackend.ts:
```diff
@@ -863,18 +863,11 @@ export const audiusBackend = ({
     )
   }
 
-  async function getAccount(fromSource = false) {
```
This was unused

audius-infra commented:
Preview this change https://demo.audius.co/rj-c-1595

sliptype commented on packages/web/src/common/store/account/sagas.js:
```diff
@@ -182,34 +181,30 @@ function* onSignedIn({ payload: { account, isSignUp = false } }) {
   yield put(accountActions.fetchAccountRequested())
 
-  if (!fromSource) {
-    yield fetchLocalAccountAsync()
```
So we don't need to call this anymore?

raymondjacobson replied:
Yeah discussed with @amendelsohn  a bit -- I can't find the utility of this b/c we fetchLocal account twice on startup currently

amendelsohn replied:
It also didn't look like `fromSource` was ever true, but possible we missed something

sliptype replied:
I couldn't find any instances of it being true either. But was worried since we are doing the local fetch in the false case. But if we are already doing fetchLocal as part of the `getAccount` logic we should be good

raymondjacobson replied:
It was never called with true lol!

sliptype commented on packages/mobile/src/screens/root-screen/RootScreen.tsx:
```diff
@@ -1,11 +1,11 @@
+import { useEffect, useState } from 'react'
 
 import { accountSelectors, Status } from '@audius/common'
+import { getHasAccount } from '@audius/common/dist/store/account/selectors'
```
I think this import is wrong, we should be destructuring from accountSelectors or something like that

raymondjacobson replied:
My bad thx

raymondjacobson commented on packages/web/src/common/store/account/sagas.js:
```diff
@@ -182,34 +181,30 @@ function* onSignedIn({ payload: { account, isSignUp = false } }) {
       })
     )
-    // Clear local storage users if present
-    yield call([localStorage, 'clearAudiusAccount'])
```
I don't think so, this causes a ton of issues though

raymondjacobson replied:
I tested pretty heavily after this change, but it makes account restore fail every other session so you're forced down the path of fetching account from remote

audius-infra commented:
Preview this change https://demo.audius.co/rj-c-1595

dylanjeffers commented on packages/mobile/src/screens/signon/SignOn.tsx:
```diff
@@ -379,8 +379,12 @@ const SignOn = ({ navigation }: SignOnProps) => {
-      setPassword('')
+
+      // Debounce resetting state
+      setTimeout(() => {
```
Do we need to cancel this in a cleanup task?

raymondjacobson replied:
i think we always want to do this, so probs not?

dylanjeffers replied:
well it's just a matter of ensuring we dont call 'setState' on an unmounted component, minor react thing, but good practice

raymondjacobson replied:
oh gotcha, will put up PR, yea

dylanjeffers reviewed (approved):
Awesome work parsing through all this complexity
